compiler/rustc_const_eval/src/check_consts/check.rs

```diff
@@ -647,55 +647,25 @@ impl<'tcx> Visitor<'tcx> for Checker<'_, 'tcx> {
-                    // typeck ensures the conditions for calling a const trait method are met,
-                    // so we only error if the trait isn't const. We try to resolve the trait
-                    // into the concrete method, and uses that for const stability checks.
-                    // FIXME(const_trait_impl) we might consider moving const stability checks
-                    // to typeck as well.
-                    if tcx.features().const_trait_impl() && trait_is_const {
-                        // This skips the check below that ensures we only call `const fn`.
-                        is_trait = true;
-
-                        if let Ok(Some(instance)) =
-                            Instance::try_resolve(tcx, param_env, callee, fn_args)
-                            && let InstanceKind::Item(def) = instance.def
-                        {
-                            // Resolve a trait method call to its concrete implementation, which may be in a
-                            // `const` trait impl. This is only used for the const stability check below, since
-                            // we want to look at the concrete impl's stability.
-                            fn_args = instance.args;
-                            callee = def;
-                        }
-                    } else {
-                        // if the trait is const but the user has not enabled the feature(s),
-                        // suggest them.
-                        let feature = if trait_is_const {
-                            Some(if tcx.features().const_trait_impl() {
-                                sym::effects
-                            } else {
-                                sym::const_trait_impl
-                            })
-                        } else {
-                            None
-                        };
-                        self.check_op(ops::FnCallNonConst {
-                            callee,
-                            args: fn_args,
-                            span: *fn_span,
-                            call_source,
-                            feature,
-                        });
-                        // If we allowed this, we're in miri-unleashed mode, so we might
-                        // as well skip the remaining checks.
-                        return;
-                    }
+                    // We are not allowed to use const traits in stable const functions yet.
+                    // FIXME: apply more fine-grained per-trait const stability checks
+                    // (see <https://github.com/rust-lang/project-const-traits/issues/5>).
+                    self.check_op(ops::FnCallNonConst {
+                        callee,
+                        args: fn_args,
+                        span: *fn_span,
+                        call_source,
+                        feature: tcx.is_const_trait(trait_did).then_some(sym::const_trait_impl),
```

Comment:
This is confusing, it conflates two entirely orthogonal checks:
- is this trait const-callable
- are we allowed to do const trait calls

Why is the first check even here, didn't `revalidate_conditional_constness` already check that?

Reply:
`revalidate_conditional_constness` just checks that if the item is *conditionally const* then the trait bounds `~const` trait bounds hold. For example, if a free function item is not conditionally const, then it will have no `~const` trait bounds, so `revalidate_conditional_constness` is not enforcing anything in that case.